flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/SerialEpochManager.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.asyncprocessing;
+
+/** An implementation of {@link AbstractEpochManager} for serial epoch execution. */
+public class SerialEpochManager extends AbstractEpochManager {
+
+    AsyncExecutionController asyncExecutionController;
+
+    public SerialEpochManager(AsyncExecutionController asyncExecutionController) {
+        super();
+        this.asyncExecutionController = asyncExecutionController;
+    }
+
+    @Override
+    public void onNonRecord(Runnable action) {
+        assert outputQueue.size() == 1;
+        Epoch lastEpoch = outputQueue.get(outputQueue.size() - 1);
+        lastEpoch.action = action;
+        lastEpoch.close();
+        asyncExecutionController.drainInflightRecords(0);
+        if (lastEpoch.tryFinish() && outputQueue.size() > 0) {
+            outputQueue.remove(0);
+        }
```

Comment:
This block of code should have been invoked in `completeOneRecord()` during `drainInFlightRecords`, thus could be removed.

Reply:
This block of code is for the case:
all records in one epoch are completed before non-records arriving, `completeOneRecord()` can't finish this epoch cause this epoch is not closed, so we need to check it again here.

Reply:
I noticed that `lastEpoch.close();` is invoked right before this block of code, so this epoch must have been closed.